### PR TITLE
Graph axis

### PR DESCRIPTION
**Pull request by local-minimum**
image

Adds optional y and x axis to bar graphs. Y axis can have ticks (both major and minor).

### CODE REVIEW

File: src/com/visualization/barchart/bar-axis-tick.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import {h, Component} 				from 'preact/preact';
+
+const MAJOR_TICK_LENGTH = 1.5;
+const MINOR_TICK_LENGTH = 0.5;
+
+export default class Tick extends Component {
+	render({major, horizontal, outside, axis, position}) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-axis-tick.js]:
```js
render( props ) {
    var {major, horizontal, outside, axis, position} = props;
```

File: src/com/visualization/barchart/bar-axis-tick.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+		pos0 = outside ? axis - (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH) : axis + (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH);
+		if (horizontal) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-axis-tick.js]:
```js
if ( horizontal ) {
```

File: src/com/visualization/barchart/bar-axis-tick.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import {h, Component} 				from 'preact/preact';
+
+const MAJOR_TICK_LENGTH = 1.5;
+const MINOR_TICK_LENGTH = 0.5;
+
+export default class Tick extends Component {
+	render({major, horizontal, outside, axis, position}) {
+		pos0 = outside ? axis - (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH) : axis + (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH);
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-axis-tick.js]:
```js
pos0 = outside ? (axis - (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH)) : (axis + (major ? MAJOR_TICK_LENGTH : MINOR_TICK_LENGTH));
```

File: src/com/visualization/barchart/bar-x-axis.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import {h, Component} 				from 'preact/preact';
+
+// Non used prop `height` is the expected maximum extent of axis
+// content down from the `yZeroPos` y.
+//
+// Component returns a group to support extension with ticks and labels
+
+export default class XAxis extends Component {
+	render({yZeroPos, padLeft, padRight}) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-x-axis.js]:
```js
render( props ) {
    var {yZeroPos, padLeft, padRight} = props;
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import {h, Component} 				from 'preact/preact';
+
+import Tick							from 'bar-axis-tick';
+
+export default class YAxis extends Component {
+	getMinorMajorTickSteps(axisMinPos, axisMaxPos, valueStep) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
getMinorMajorTickSteps( axisMinPos, axisMaxPos, valueStep ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		const valueRangeLength = (axisMaxPos - axisMinPos) / valueStep;
+		let majorTick = 1;
+		let minorTick = 1;
+		if (valueRangeLength > 100) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
if ( valueRangeLength > 100 ) {
```
(and others below like this)

File: src/com/visualization/barchart/bar-y-axis.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			minorTick = 5;
+		}
+		else if (valueRangeLength > 10) {
+			majorTick = 5;
+			minorTick = 1;
+		}
+		return {'minorTick': minorTick, 'majorTick': majorTick};
+	}
+
+	render({yZeroPos, yOnePos, showTicks, padBottom, padTop, styleClass, width}) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
render( props ) {
   var {yZeroPos, yOnePos, showTicks, padBottom, padTop, styleClass, width} = props;
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		let ShowTicks = null;
+		if (yZeroPos !== null && yOnePos !== null && showTicks == true) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
if ( (yZeroPos !== null) && (yOnePos !== null) && (showTicks == true) ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+			if (showTicks === true) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
if ( showTicks === true ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+				ShowTicks = [];
+				const step = yOnePos - yZeroPos;
+				const axisMaxPos = 100 - padTop;
+				const {minorTick, majorTick} = this.getMinorMajorTickSteps(padBottom, axisMaxPos, step);
+
+				let tick = 0;
+				let tickHeight = yZeroPos;
+				while (true) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
while ( true ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+					tick += minorTick;
+					tickHeight = yZeroPos + (tick * step);
+					if (tickHeight >= axisMaxPos) {
+						break;
+					}
+					if (tick % majorTick === 0) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
if ( (tick % majorTick) === 0 ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+				const step = yOnePos - yZeroPos;
+				const axisMaxPos = 100 - padTop;
+				const {minorTick, majorTick} = this.getMinorMajorTickSteps(padBottom, axisMaxPos, step);
+
+				let tick = 0;
+				let tickHeight = yZeroPos;
+				while (true) {
+					tick += minorTick;
+					tickHeight = yZeroPos + (tick * step);
+					if (tickHeight >= axisMaxPos) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
if ( tickHeight >= axisMaxPos ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+						break;
+					}
+					if (tick % majorTick === 0) {
+						ShowTicks.push(<Tick major horizontal outside axis={width} position={tickHeight} />);
+					}
+					else if (tick % majorTick) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
else if ( tick % majorTick ) {
```

File: src/com/visualization/barchart/bar-y-axis.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+						ShowTicks.push(<Tick minor horizontal outside axis={width} position={tickHeight} />);
+					}
+				}
+
+				tickHeight = yZeroPos;
+				tick = 0;
+				while (true) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar-y-axis.js]:
```js
while ( true ) {
```
(and the if's below)

File: src/com/visualization/barchart/bar.js [hunk 15]
```diff
@@ -1,37 +1,31 @@
...
+
 export default class Bar extends Component {
 
 	constructor( props ) {
 		super(props);
 
     }
 
-    render( props ) {
+    render( {valuePos, zero, width, index, color, left} ) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar.js]:
```js
 render( props ) {
   var {valuePos, zero, width, index, color, left} = props;
```

File: src/com/visualization/barchart/bar.js [hunk 16]
```diff
@@ -1,37 +1,31 @@
...
 
-        // it's valid for height/width/index to be zero so we have to check it against undefined.
-        if ( !(props && (props.height != null) && (props.index != null) && (props.width != null) && props.color) ) {
-            console.warn('Bar was created with invalid props', props);
+        // it's valid for valuePos/width/index to be zero so we have to check it against undefined.
+        if ( !(valuePos != null && index != null && width != null && left != null && color )) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar.js]:
```js
if ( !((valuePos != null) && (index != null) && (width != null) && (left != null) && color) ) {
```

File: src/com/visualization/barchart/bar.js [hunk 17]
```diff
@@ -1,37 +1,31 @@
...
+            console.warn('Bar was created with invalid props', this.props);
             return;
         }
 
-        let {height, zero, width, index, color} = props;
-		if (height < 0) {
+		if (valuePos < zero) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/bar.js]:
```js
if ( valuePos < zero ) {
```

File: src/com/visualization/barchart/barchart.js [hunk 18]
```diff
@@ -2,54 +2,97 @@ import {h, Component} 				from 'preact/preact';
...
+		let ShowYAxis = null;
+		if (props.showYAxis) {
+			ShowYAxis = <YAxis yZeroPos={yZeroPos} yOnePos={yOnePos} xZeroPos={xZeroPos} padTop={padTop} padBottom={padBottom} width={yAxisWidth} showTicks={props.showYTicks} />;
+		}
+
+		let barWidth = this.scale(1, xMaxValue, xZeroPos, 100 - padRight) - xZeroPos;
 
 		let total = values.reduce((a, b) => a + b, 0);
-		let percentages = values.map((x) => { return Math.round((100*(x/total))*100)/100; });
+		let percentages = values.map(x => Math.round((100*(x/total))*100)/100);
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/barchart.js]:
```js
let percentages = values.map((x) => (Math.round((100*(x/total))*100)/100));
```

File: src/com/visualization/barchart/barchart.js [hunk 19]
```diff
@@ -2,54 +2,97 @@ import {h, Component} 				from 'preact/preact';
...
 
 		let Bars = [];
         let Names = [];
 		let Colors = [];
 
 		let ShowLegend = null;
-		for ( var i = 0; i < values.length; i++ ) {
+		for ( let i = 0; i < values.length; i++ ) {
+
+			if (valuesYPos[i] == yZeroPos || isNaN(valuesYPos[i])) {
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/barchart.js]:
```js
if ( (valuesYPos[i] == yZeroPos) || isNaN(valuesYPos[i]) ) {
```

File: src/com/visualization/barchart/barchart.js [hunk 20]
```diff
@@ -2,54 +2,97 @@ import {h, Component} 				from 'preact/preact';
...
+				//continue;
+			}
 
 			let color = 1 + ( i % 6 );
```
Comment (mikekasprzak) [inline src/com/visualization/barchart/barchart.js]:
```js
let color = 1 + (i % 6);
```

**Comment by mikekasprzak** [comment]
See #1569 for details on the style notes.

File: src/com/visualization/barchart/bar-axis-tick.js (hunk 1 above)
Comment (local-minimum) [reply src/com/visualization/barchart/bar-axis-tick.js]: I'm going to hope I can still use `let` and `const` right?

**Pull request merged**